git-meta.1
'\" t
.\"     Title: git-meta
.\"    Author: [see the "AUTHORS" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 01/04/2016
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "GIT\-META" "1" "01/04/2016" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
git-meta \- Handle file meta information within git
.SH "SYNOPSIS"
.sp
git\-meta [ options ] command
.SH "DESCRIPTION"
.sp
Git (currently) cannot handle file system meta information such as permissions, last modification time or ownership (both user and group)\&. Here comes git\-meta to fix this in some cases where these informations are mandatory or simply useful\&.
.sp
All informations are stored in a \&.gitmeta file managed itself by git\&.
.sp
Using git hooks may automate the backup and restore process\&.
.SH "OPTIONS"
.PP
\-\-data DATA
.RS 4
Set the
git\-meta
database file (default:
\fI\&.gitmeta\fR)\&.
.RE
.PP
\-f|\-\-force
.RS 4
Force writing data even if no modification has been done in the git tree\&. By default
git\-meta
does not run if no modification has been done to prevent from massive git change set creation\&.
.RE
.PP
\-h|\-\-help
.RS 4
Display help screen\&.
.RE
.PP
\-a|\-\-add\-to\-git
.RS 4
Add database to the git commit list\&.
.RE
.PP
\-O|\-\-owner
.RS 4
Store symbolic file ownership to the database\&.
.RE
.PP
\-o|\-\-numeric\-owner
.RS 4
Store numeric file ownership to the database\&.
.RE
.PP
\-P|\-\-skip\-perms
.RS 4
Skip file perms during restore\&.
.RE
.PP
\-M|\-\-skip\-mtime
.RS 4
Skip file mtime during store/restore operations\&. The mtime is not written into the database\&.
.RE
.PP
\-U|\-\-skip\-user
.RS 4
Skip file owner during restore\&.
.RE
.PP
\-G|\-\-skip\-group
.RS 4
Skip group owner during restore\&.
.RE
.PP
\-i|\-\-ignore BRANCH
.RS 4
Branch on witch
git\-meta
is ignored\&. This is treated as a regexp\&. This option can be specified several times\&. if not specfied, values are search in the
meta\&.ignore
git configuration key\&.
.RE
.SH "COMMANDS"
.PP
get
.RS 4
Store metadata to database\&.
.RE
.PP
set
.RS 4
Restore metadata from database\&.
.RE
.PP
dump
.RS 4
Dump metadata from database\&.
.RE
.PP
init
.RS 4
Create hooks in git hooks directory\&.
.RE
.SH "DATABASE FORMAT"
.sp
The database format is quiet simple\&. It consists of a JSON dump of a dictionnary for which the file name is the key and the file attributes are the values\&. dictionnary\&.
.SS "Dictionnary format"
.sp
The key is composed of the full filename path starting from the git repository root\&.
.sp
Each values are:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
mode: the file permissions in decimal format\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
mtime: the file mtime in seconds since Epoch\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
uid: the file owner (in numerical or symbolic format)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
gid: the file group (in numerical or symbolic format)\&.
.RE
.SS "File ownership"
.sp
File ownership (both owner and group) are either:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Not stored at all if neighter the
\fI\-o\fR
nor
\fI\-O\fR
option is given\&. This might be used if the files ownership may change from one system to an other\&. Generally this should be used in almost all cases\&. Otherwise a large amount of changesets could pollute git commits\&. In that case the last 2 fields (both owner and group) of each record are not present in the database\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Stored in symbolic format if
\fI\-O\fR
is given\&. This option should be used if symbolic ownership are the same in all systems (for keeping track of
\fI/etc\fR
changes for example)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Stored in numerical format if
\fI\-o\fR
is given\&. This option should be used if numerical ownerships are the same in all systems\&.
.RE
.SH "USING WITH GIT HOOKS"
.sp
Manual use git\-meta is not that much handy\&. It can be configured to be used directly withing githooks(5), specially \fBpre\-commit\fR for the backup and \fBpost\-merge\fR for the restore\&.
.SS "pre\-commit example"
.sp
.if n \{\
.RS 4
.\}
.nf
#!/bin/sh

git meta \-a get || true
.fi
.if n \{\
.RE
.\}
.SS "post\-merge example"
.sp
.if n \{\
.RS 4
.\}
.nf
#!/bin/sh

git meta set || true
.fi
.if n \{\
.RE
.\}
.SH "SEE ALSO"
.sp
git(1), githooks(5)\&.
.SH "HISTORY"
.PP
2014/04/24
.RS 4
Add ignore option
.RE
.PP
2014/03/02
.RS 4
Rewite script in python
.RE
.PP
2011/09/11
.RS 4
Add init command\&.
.RE
.PP
2011/03/26
.RS 4
First public release\&.
.RE
.SH "BUGS"
.sp
No time to include bugs, command actions might seldom lead astray user\(cqs assumption\&.
.SH "AUTHORS"
.sp
git\-meta is written by S\('ebastien Gross <seb\(buɑƬ\(buchezwam\(buɖɵʈ\(buorg>\&.
.SH "COPYRIGHT"
.sp
Copyright \(co 2011\-2014 S\('ebastien Gross <seb\(buɑƬ\(buchezwam\(buɖɵʈ\(buorg>\&.
.sp
Relased under WTFPL (\m[blue]\fBhttp://sam\&.zoy\&.org/wtfpl/COPYING\fR\m[])\&.


debian/changelog
git-meta (2.0-1) unstable; urgency=low

  * Bump to version 2.

 -- Sébastien Gross <seb•ɑƬ•chezwam•ɖɵʈ•org>  Mon, 03 Mar 2014 14:28:41 +0000

git-meta (0.1-1) unstable; urgency=low

  * Initial release.

 -- Sébastien Gross <seb•ɑƬ•chezwam•ɖɵʈ•org>  Sat, 26 Mar 2011 09:38:11 +0000



git-meta.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

"""
# git-meta --- Handle file metadata in git

"""

# Copyright © 2014 Sébastien Gross <seb•ɑƬ•chezwam•ɖɵʈ•org>
# Created: 2014-02-26
# Last changed: 2014-03-03 14:26:03

# This program is free software. It comes without any warranty, to
# the extent permitted by applicable law. You can redistribute it
# and/or modify it under the terms of the Do What The Fuck You Want
# To Public License, Version 2, as published by Sam Hocevar. See
# http://sam.zoy.org/wtfpl/COPYING for more details.


__author__ = "Sébastien Gross"
__copyright__ = """Copyright © 2014 Sébastien Gross <seb•ɑƬ•chezwam•ɖɵʈ•org>"""

import os, sys, subprocess
import pipes
from stat import *
import pwd
import grp
import json
import argparse
import pprint
import datetime

def run(cmd):
    """run cmd"""
    proc = subprocess.Popen(cmd,
                            stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE)
    proc.wait()
    out, err = proc.communicate()
    rc = proc.returncode
    if rc == 0:
        return out
    else:
        sys.stdout.write("Failed (%s): %s\n" %
                         (rc, ' '.join(pipes.quote(arg) for arg in cmd)))
        sys.stderr.write(err)
        sys.exit(1)

def get_file_stats(f, args):
    st = os.stat(f)
    ret = { 'mode': st.st_mode, 'mtime': st.st_mtime }
    if args['numeric_owner'] is True:
        ret['uid'] = st.st_uid
        ret['gid'] = st.st_gid
    elif args['owner'] is True:
        ret['uid'] = pwd.getpwuid(st.st_uid).pw_name,
        ret['gid'] =  grp.getgrgid(st.st_gid).gr_name
    return ret
    

def git_get_files_attributs(args):
    """ """
    out = run(['git', 'ls-files', '-z'])
    files=out.split('\0')
    files.sort()
    rows = {}
    for f in files:
        if not(os.path.isfile(f)):
            continue
        rows[f] = get_file_stats(f, args)
        while not (f == '' or f == '.'):
            if not rows.has_key(f):
                rows[f] = get_file_stats(f, args)
            f = os.path.dirname(f)
    metadata = open(args['data'], 'w')
    json.dump(rows, metadata, indent=0)
    metadata.close()

def load_metadata(args):
    """Load metada from file"""
    p = open(args['data'], 'r')
    ret = json.load(p)
    p.close()
    return ret

def dump_database(args):
    data = load_metadata(args)
    entries = data.keys()
    entries.sort()
    print('%s %s %s %s %s' %( 'mode'.center(6), 'mtime'.center(19), 'uid'.center(6), 'gid'.center(6), 'file' ))
    for entry in entries:
        row = ['%.6o' % data[entry]['mode']]
        if data[entry].has_key('mtime'):
            row.append(datetime.datetime.fromtimestamp(data[entry]['mtime']).strftime('%Y-%m-%d %H:%M:%S'))
        if data[entry].has_key('uid'):
            row.append('%6s' % data[entry]['uid'])
        if data[entry].has_key('gid'):
            row.append('%6s' % data[entry]['gid'])
        if os.path.isfile(entry):
            row.append(entry)
        else:
            row.append('%s/' % entry)
        print ' '.join(row)

def git_set_file_attributs(args):
    data = load_metadata(args)
    for file in data.keys():
        if args['verbose']: sys.stdout.write('%s:' % file)
        if (args['skip_perms'] is False) or not(data[file].has_key('mode')):
            if args['verbose']: sys.stdout.write(' mode:%6o' % data[file]['mode'])
            os.chmod(file, data[file]['mode'])
        if (args['skip_mtime'] is False)  or not(data[file].has_key('mtime')):
            if args['verbose']: sys.stdout.write(' mtime:%s' % datetime.datetime.fromtimestamp(data[file]['mtime']).strftime('%Y-%m-%d %H:%M:%S'))
            os.utime(file, (data[file]['mtime'], data[file]['mtime']))
        uid = -1
        gid = -1
        if (args['skip_user'] is False) or not(data[file].has_key('uid')) or \
           (args['skip_group'] is False) or not(data[file].has_key('gid')):
            if (args['skip_user'] is False) or not(data[file].has_key('uid')):
                if isinstance(data[file]['uid'], int):
                    uid = data[file]['uid']
                else:
                    uid = pwd.getpwnam(data[file]['uid']).pw_uid
            if (args['skip_group'] is False) or not(data[file].has_key('gid')):
                if isinstance(data[file]['gid'], int):
                    gid = data[file]['gid']
                else:
                    gid = grp.getgrnam(data[file]['gid']).gr_gid
            if args['verbose']: sys.stdout.write(' user:%s group:%s' % (uid, gid))
            os.chown(file, uid, gid)
        if args['verbose']: sys.stdout.write('\n')


def parse_cmd_line():
    """Parse command line arguments"""
    args = argparse.ArgumentParser(description="Command line parser")

    args.add_argument('--data', help='Path to database', metavar='PATH',
                      default='.gitmeta')
    args.add_argument('-f', '--force', help='Force to commit',
                      default=False, action='store_true')
    args.add_argument('-O', '--owner', help='Add symbolic owner',
                      default=False, action='store_true')
    args.add_argument('-o', '--numeric-owner', help='Add numeric owner',
                      default=False, action='store_true')
    args.add_argument('-a', '--add-to-git', help='Commit database to git',
                      default=False, action='store_true')
    args.add_argument('-U', '--skip-user', help='Skip user during restore',
                      default=False, action='store_true')
    args.add_argument('-G', '--skip-group', help='Skip group during restore',
                      default=False, action='store_true')
    args.add_argument('-P', '--skip-perms', help='Skip perms during restore',
                      default=False, action='store_true')
    args.add_argument('-M', '--skip-mtime', help='Skip mtime during restore',
                      default=False, action='store_true')
    args.add_argument('-v', '--verbose', help='Show operations',
                      default=False, action='store_true')


    sp_cmd = args.add_subparsers(help='Command', dest='cmd')

    sp_cmd.add_parser('init', help='Setup hooks in .git directory')
    sp_cmd.add_parser('get', help='Store metadata')
    sp_cmd.add_parser('set', help='Restore metadata')
    sp_cmd.add_parser('dump', help='Dump metadata')

    return args.parse_args().__dict__


def git_init_hooks():
    """Initialize git hooks"""
    out = run(['git', 'rev-parse', '--git-dir'])
    dir = out.split('\n')[0]
    f = open('%s/hooks/pre-commit' % dir, 'w')
    f.write( "#!/bin/sh\n\n"
    "git meta -a get || true\n")
    f.close()
    f = open('%s/hooks/post-merge' % dir, 'w')
    f.write( "#!/bin/sh\n\n"
    "git meta -a set || true\n")
    f.close()

def __init__():
    args = parse_cmd_line()
    if args['cmd'] == 'get':
        git_get_files_attributs(args)
    elif args['cmd'] == 'dump':
        dump_database(args)
    elif args['cmd'] == 'set':
        git_set_file_attributs(args)
    elif args['cmd'] == 'init':
        git_init_hooks()
if __name__ == "__main__":
    __init__()
